### DISCUSSION TITLE
TypeError: Failed to fetch

### DISCUSSION THREAD

**Discussion by brunohmsato** [balta-io/forum#659] [Tira Dúvidas] [answered]
Boa tarde Balta. Acompanhando os vídeos do curso FullStack .NET e cheguei no momento de cadastrar o usuário pela aplicação, porém fica dando erro e é exibido a seguinte mensagem --> "TypeError: Failed to fetch". 

Tentei fazer o cadastro diretamente via o swagger, e executou normalmente. No visual studio 2022, tenho o seguinte retorno mais detalhado. 

System.Net.Http.HttpRequestException: TypeError: Failed to fetch
 ---> TypeError: Failed to fetch
   --- End of inner exception stack trace ---
   at System.Net.Http.BrowserHttpInterop.<CancelationHelper>d__13`1[[System.Runtime.InteropServices.JavaScript.JSObject, System.Runtime.InteropServices.JavaScript, Version=8.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]].MoveNext()
   at System.Net.Http.BrowserHttpHandler.CallFetch(HttpRequestMessage request, CancellationToken cancellationToken, Nullable`1 allowAutoRedirect)
   at System.Net.Http.BrowserHttpHandler.<SendAsync>g__Impl|53_0(HttpRequestMessage request, CancellationToken cancellationToken, Nullable`1 allowAutoRedirect)
   at Microsoft.Extensions.Http.Logging.LoggingHttpMessageHandler.<SendCoreAsync>g__Core|5_0(HttpRequestMessage request, Boolean useAsync, CancellationToken cancellationToken)
   at Microsoft.Extensions.Http.Logging.LoggingScopeHttpMessageHandler.<SendCoreAsync>g__Core|5_0(HttpRequestMessage request, Boolean useAsync, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.<SendAsync>g__Core|83_0(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationTokenSource cts, Boolean disposeCts, CancellationTokenSource pendingRequestsCts, CancellationToken originalCancellationToken)
   at Dima.Web.Handlers.AccountHandler.RegisterAsync(RegisterRequest request) in C:\Users\bruno.sato.FENOXNET\source\repos\Dima\Dima.Web\Handlers\AccountHandler.cs:line 23
   at Dima.Web.Pages.Identity.RegisterPage.OnValidSubmitAsync() in C:\Users\bruno.sato.FENOXNET\source\repos\Dima\Dima.Web\Pages\Identity\Register.razor.cs:line 54

**Comment by andrebaltieri** [comment, marked as answer]
Bom dia, @brunohmsato, como vai?

Posta os trechos de código destes arquivos aqui:
```
C:\Users\bruno.sato.FENOXNET\source\repos\Dima\Dima.Web\Handlers\AccountHandler.cs:line 23
at Dima.Web.Pages.Identity.RegisterPage.OnValidSubmitAsync() in C:\Users\bruno.sato.FENOXNET\source\repos\Dima\Dima.Web\Pages\Identity\Register.razor.cs:line 54
```

Se tiver um print ou algo pra ajudar...

**Reply by brunohmsato** [reply]
Bom dia. Claro, seguem os trechos:

![Screenshot_1](https://github.com/user-attachments/assets/7f791024-0f7d-4053-ab7d-c7754344b2d3)
![Screenshot_2](https://github.com/user-attachments/assets/c023825b-95d6-4d55-b105-931217df68f9)
![Screenshot_3](https://github.com/user-attachments/assets/3af37656-c370-4907-b654-309373d8b059)

**Reply by andrebaltieri** [reply]
Bom dia, @brunohmsato, como vai?

Coloca um breakpoint na linha 23 do seu AccountHandler por favor, veja os detalhes do erro que dá e manda um print aqui.
Aparentemente o Post ocorreu, mas o retorno não foi possível ser traduzido.

Pode ser que o post ocorreu com sucesso, mas retornou dados inesperados... 
O request está correto? Usuário e senha corretos? Os mesmos que usou no teste da API?

**Reply by brunohmsato** [reply]
Boa tarde Balta, tudo certo sim.

Coloco o breakpoint no RegisterAsync, porém estando na linha 23 e executando f11, ele simplesmente continua e não exibe mais informações.

Tenho conferido no banco de dados, mas nenhum registro é adicionado à tabela.

**Reply by andrebaltieri** [reply]
Tentou um build/clean/rebuild? Está bem estranho este comportamento... aparentemente seu código está OK!
Minha dúvida era mais se ele estava mandando para o método certo.

**Reply by brunohmsato** [reply]
Bom dia! SIm, tentei todos esses procedimentos, e recapitulei os vídeos para ver se não tinha deixado passar nada. 

Bom, independente... vou dar continuidade e caso consiga descobrir o que pode estar acontecendo, faço o envio aqui. E muito obrigado 
pela atenção!

**Comment by thaiguer** [comment]
Olá, @brunohmsato , @andrebaltieri.
Me deparei com o mesmo problema.

### TLDR:
Adicionando o a linha 23, consegui dar continuidade com sucesso ao cadastro de usuário:
![image](https://github.com/user-attachments/assets/0b65587a-92fc-4d00-808b-22137d23225d)

### Continuando:
Não sei o motivo, mas a propriedade BaseAddress, do _client, não ficava setada para o "https://localhost:7119/" e sim "http://localhost:7119/" (sem o 's').

Mesmo estando com o s no arquivo do configuração:
```chasrp public static string BackendUrl { get; set; } = "https://localhost:7119";```

### Durante o debug:
![image](https://github.com/user-attachments/assets/2af9c3c6-ff88-467b-bf84-2df438a68682)

### Atribuição da propriedade BackendUrl:
![image](https://github.com/user-attachments/assets/60464df7-5df3-4a97-a8d2-1add12c1eb58)

Valores no Console:
![image](https://github.com/user-attachments/assets/31d74a97-474a-422d-a848-fcd1d53cb476)
Nota: Printei no Console porque ao colocar o mouse sobre a propriedade o valor não aparece.

### Concluindo:
Não parece ser um problema claro no código, pois ao clonar o repositório do Balta, também ocorre o mesmo erro no meu PC.
Mas há algum problema ocorrendo no set da linha 12 do program.cs do Dima.Web.
Tenho a NordVpn, talvez o software cause confusão nos dispositivos de rede do Windows.

**Comment by LeandroCafe1** [comment]
Olá, @brunohmsato, @andrebaltieri e @thaiguer  Boa noite!

Me deparei com o mesmo erro que vocês relataram, com a mensagem: 'TypeError: Failed to fetch', ao tentar realizar o login ou cadastro.

Tentei identificar a causa, mas não obtive sucesso.  cheguei nos mesmos pontos relatados acima. 

Há alguma novidade sobre o problema? Conseguiram corrigir esse erro?

**Reply by thaiguer** [reply]
Dei continuidade ao curso forçando o valor do BaseAdress (hard coded). Mesmo trocando de sistema (Ubuntu Desktop), formatando o PC (Windows 11) o mesmo se repetiu para mim. Clonando o projeto do Balta também obtive esse erro. Não investiguei o problema mais a fundo, mas não parece ser por causa da VPN no meu caso, pois formatando o PC e não instalando a VPN o problema persistiu.

**Reply by fjlourenco** [reply]
Olá @LeandroCafe1,

Também verifiquei a mesma ocorrência e, conforme o @thaiguer havia apontado acima, no meu caso havia também uma certa confusão entre protocolos HTTP e HTTPS. Inicialmente verifiquei se havia algum tipo de restrição quanto ao CORS. Usei a aba Network do DevTools para isso. Após ajustes no CORS, o problema continuou e estava relacionado a um redirecionamento que o pacote AspNetCore.Identity estava fazendo. Mesmo configurando a URL do Backend, ele ainda insistia em redirecionar internamente do protocolo HTTPS para o protocolo (com respectiva porta) HTTP. Para corrigir o problema que encontrei aqui, alterei a ordem dos perfis que estão no **Dima.Web\launchSettings.js**, onde coloquei o perfil HTTPS como o primeiro item, conforme figura abaixo:
![image](https://github.com/user-attachments/assets/3afbcbbd-e291-4e04-9863-d26d66cfd3db)

Desta forma, tanto fazendo o debug pelo VS Code quanto iniciando a API e o projeto WEB por linha de comando (usando o dotnet run e dotnet  watch run no caso do Dima.Web) funcionaram. Como acabei utilizando HTTPS ao invés do HTTP, que foi o protocolo utilizado no curso, acabei me deparando com esta questão específica. Não sei se este é exatamento o teu caso, mas deixo aqui algo que para mim funcionou.
Abraço!

**Reply by LeandroCafe1** [reply]
Olá, @fjlourenco, @thaiguer, @brunohmsato , @andrebaltieri.

Primeiramente, gostaria de agradecer pelas orientações. Realizei os ajustes conforme indicado, porém o erro persiste:

segue abaixo mais detalhes.

![image](https://github.com/user-attachments/assets/977a0a91-3f68-456a-b52f-6513bb333ea5)
![image](https://github.com/user-attachments/assets/c7c22e28-c39b-4376-883d-4c1639f00990)
![image](https://github.com/user-attachments/assets/78ad2644-741e-4213-812c-c6a6ca020db9)
![image](https://github.com/user-attachments/assets/41d14f06-a123-4f03-94fb-7e254e8b3c51)

![image](https://github.com/user-attachments/assets/0aa0ba2c-0318-4493-9bef-61ca2c43bcec)

Infelizmente, ainda não funcionou. Ainda não encontrei o que pode estar provocando este erro.

Sabe se preciso fazer mais algum ajuste ?


Obrigado

**Reply by thaiguer** [reply]
Imagino que pelo swagger esteja tudo funcionando. Certo?
Inspeciona qual é a URL que o frontend está gerando para fazer a requisição na API.

**Reply by fjlourenco** [reply]
Sobe o projeto aqui no Github.

**Comment by LeandroCafe1** [comment]
Bom dia, pessoal!

Identifiquei a causa do problema e já foi resolvido. Tudo está funcionando perfeitamente agora.

Obrigado pelo apoio!

**Comment by thaiguer** [comment]
> Bom dia, pessoal!
> 
> Identifiquei a causa do problema e já foi resolvido. Tudo está funcionando perfeitamente agora.
> 
> Obrigado pelo apoio!

Se puder deixa uma breve explicação de como, ou das alterações que você fez no projeto.

**Reply by LeandroCafe1** [reply]
Para resolver o problema, foi necessário ajustar a porta e reorganizar a ordem dos perfis no arquivo launchSettings.js.

**Reply by fjlourenco** [reply]
Show!!!
